### ISSUE TITLE
[macOS] Push-to-talk hotkey for audio sources does not seem to function correctly 

### ISSUE THREAD

**Issue by jcm93** [obsproject/obs-studio#9023]
### Operating System Info

macOS 13

### OBS Studio Version

29.1.2

### OBS Studio Log URL

https://obsproject.com/logs/fUBR7DZXd4NgTuI8

### Expected Behavior

Push to talk should operate correctly, i.e. muting the audio source unless the selected button is currently pressed down.

### Current Behavior

I've seen different behavior. Currently, on my system, the push-to-talk hotkey seems to have no effect whatsoever. It neither mutes nor unmutes the audio source.

In the past, I've seen it momentarily engage the audio source for a fraction of a second and then immediately mute it even if the button is still pressed, but that behavior doesn't seem to be present now.

### Steps to Reproduce

1. Add a source with audio (for my testing, I used an Audio Input Source associated with my built-in microphone).
2. In Settings->Hotkeys, set a push-to-talk key for the audio source.
3. With or without OBS having focus, press the hotkey; observe no response in the audio mixer or in the output.

### Anything else we should know?

I've made sure OBS has Accessibility permissions as well as Input Monitoring permissions. I've tried removing those permissions and re-granting them, but it does not seem to have any effect.

The mute and unmute hotkeys seem to work correctly. Push-to-mute, however, also seems similarly broken.

Also: I believe that push-to-talk has not functioned correctly for some time; I *believe* at least as far back as 27.2.4 (though the issue presented slightly differently).

**Comment by PatTheMav** [comment]
Preliminary investigation shows that the functionality works as intended _as long as the OBS window is not focused_. In that situation the key events are handled primarily by macOS itself, but when the window is in focus we receive the key events via Qt's internal key event system.

And in the latter case it seems like when pressing (and holding) the key combination we get a single event for the pressed keys, but after that the event is consumed and our hotkey system assumes that the keys are not pressed anymore, hence why "push to talk" only works for quick moment.

**Comment by jcm93** [comment]
I am still having issues here, seeing behavior that is different from what Pat described. Even with Pat's PR, "push-to" hotkeys don't seem to do anything on my system, regardless of whether OBS is focused or not. Have not had time to investigate in detail yet.

**Comment by PatTheMav** [comment]
> I am still having issues here, seeing behavior that is different from what Pat described. Even with Pat's PR, "push-to" hotkeys don't seem to do anything on my system, regardless of whether OBS is focused or not. Have not had time to investigate in detail yet.

Did you grant Accessibility access permissions to OBS Studio?

**Comment by jcm93** [comment]
Yes, made sure accessibility is granted. My own preliminary investigation is showing that, with OBS not in focus, the events are not even reaching the handler in ``obs-cocoa.m`` for my hotkeys, though they are reaching the handler for modifier keys such as Command. I was about to debug my responder chain but confirmed with user Feets in Discord that he is experiencing the same, or at least similar, behavior.

**Comment by jcm93** [comment]
I'm now seeing behavior that contradicts my last post. Summarizing the current behavior with master:

With OBS in focus:

"Mute" and "Unmute" hotkeys work correctly.
"Push-to-mute" and "Push-to-talk" hotkeys seem to do nothing. The key events themselves are picked up by ``OBSBasic::HotkeyTriggered``, but they are not functioning. Have not investigated why yet.

With OBS not in focus:

No hotkeys function correctly. The events are reaching ``obs-cocoa.m``'s ``handle_monitor_event``. They are not, however, producing any functionality. Have not investigated why yet.

**Comment by jcm93** [comment]
https://github.com/obsproject/obs-studio/assets/6864788/5a6f5b81-a266-487b-b192-f8c78d90c782

This is the kind of silliness I am seeing. Looks like a multithreading issue. Sometimes out of focus, hotkeys work, sometimes they don't.

(video summary: Hotkeys working in OBS; tab to Finder, hotkeys don't work. Tab back into OBS, hotkeys work, tab back to Finder, hotkeys now work).

**Comment by PatTheMav** [comment]
> obshotkeys2.mp4 
> 
> This is the kind of silliness I am seeing. Looks like a multithreading issue. Sometimes out of focus, hotkeys work, sometimes they don't.
> 
> (video summary: Hotkeys working in OBS; tab to Finder, hotkeys don't work. Tab back into OBS, hotkeys work, tab back to Finder, hotkeys now work).

I can't make out anything in that video - you probably might want to slow your actions down by _a lot_ and also annotate/highlight/freeze frame things in it. Otherwise a video is not the best way to explain the issue.

Note that you have to specifically enable "Push to Talk"/"Push to Mute" for a source in the audio settings, otherwise the hotkey will not do anything.

**Comment by jcm93** [comment]
> I can't make out anything in that video - you probably might want to slow your actions down by _a lot_ and also annotate/highlight/freeze frame things in it. Otherwise a video is not the best way to explain the issue.
> 
> Note that you have to specifically enable "Push to Talk"/"Push to Mute" for a source in the audio settings, otherwise the hotkey will not do anything.

The video shows that, generally, hotkeys are not working, regardless of push to talk or push to mute. Sorry to abruptly pivot away from push-to-talk specifically, but I figured it makes more sense to debug hotkeys generally not responding first. I can open a separate issue or adjust the title of this issue if that would make more sense.

At 0:05, the video shows that the hotkey to mute my audio source is `E`, and the hotkey to unmute it is `]`.

At 0:07 - 0:09, as shown on the Accessibility Keyboard Viewer, I press the hotkeys several times with OBS in focus. The source mutes and unmutes as expected.

At 0:11, I navigate to Finder, and press the hotkeys several times once more. There is no response from the source in the audio mixer.

At 0:17, I move back to OBS, and press the hotkeys several times again. The source mutes and unmutes as expected.

At 0:19, I switch to Finder once more, and press the hotkeys several times again. This time, the source now mutes and unmutes as expected.

**Comment by PatTheMav** [comment]
FWIW I just tested all these things and they work as expected:

* Mute and Unmute with OBS in focus
* Mute and Unmute with Finder in focus
* "Push to Talk" with OBS in focus
* "Push to Talk" with Finder in Focus

I used single keystrokes (`E` and `,` respectively) and modifier keystrokes with up to and including 4 modifier keys. Build was based on current master branch, all required permissions were granted (no input monitoring, as it's not needed).

**Comment by jcm93** [comment]
Also for what it's worth, this is the behavior of push-to-talk on my system. It does not do anything, regardless of whether OBS is in focus or not. Current master. Possible this is even a separate issue.

https://github.com/obsproject/obs-studio/assets/6864788/79330017-943e-4c01-b165-b176618d77be

https://github.com/obsproject/obs-studio/assets/6864788/56ff73ea-ab73-4bc4-afaa-9be75dd4f35c

What SoC are you using? I am on the M1 Max. If there is a multithreading issue, maybe it is only manifesting on some systems.

**Comment by PatTheMav** [comment]
I'm also on M1 Max, but I just tested it on M2 Pro and it works as well.

The app needs to be properly codesigned for this to work, and the entry in the accessibility access list needs to represent this codesigned variant of the app (if you're not using a stable codesign identity but instead use "sign to run on my machine" you have to remove the entry between code changes). After granting access and relaunching the app it should work.

**Comment by jcm93** [comment]
For all of the above tests, I have had code signing set up under my personal team and working properly. Before I set up codesigning and added Accessibility permissions, I removed the app from the permissions lists in System Settings as well as ran ``tccutil`` to reset all entries for OBS (in my tests, this was necessary for the Accessibility permission to properly reset). Then I launched OBS via Xcode, re-added the relevant permissions, and confirmed that OBS still had permissions between code changes.

As far as I know, if OBS did not have Accessibility permissions, it wouldn't be possible for it to *ever* pick up my hotkeys with OBS in focus, while, in the first video I posted in this thread, you can clearly see that it *sometimes* picks up my hotkeys with OBS not in focus.

**Comment by jcm93** [comment]
In my tests, I assumed that Accessibility was granted properly if, on keypress, with OBS out of focus, ``handle_monitor_event`` was reached in `obs-cocoa.m`. Even in cases where hotkeys were not working, this function would be reached on keypress. Adding the breakpoint and entering the debugger, however, would cause the hotkey to work. The "hotkeys generally not functioning" behavior, then, would only be seen if I was not setting an active breakpoint in that function.

**Comment by PatTheMav** [comment]
IIRC `NSEventMaskKeyDown` has stricter requirements than `NSEventTypeFlagsChanged`, because if you set a breakpoint in that conditional it should be hit even without proper permissions, but getting actual key events requires the entire chain to be correct:

* Application is codesigned
* Accessibility permissions are granted to application
* Application signature at the time of granting the permission is _identical_ to signature of running application

If you codesign the app without an Apple Developer account, its code signature will be different every time a compilation unit changes (the application is identified by it's `cdhash` instead of the developer certificate). So even if you granted the permission once, it will not apply to the newly compiled variant you're trying to run.

**Comment by jcm93** [comment]
I have a (paid) Apple Developer account that I am signing with, using the "Mac Developer" signing certificate, and I am confident the signing is set up correctly. I can break specifically on ``plat->is_key_down[obs_key_from_virtual_key(event.keyCode)] = (event.type == NSEventTypeKeyDown);``

This line is reached. The hotkey does not, however, do anything, approximately half the time. The behavior is inconsistent, but I will happily record a video demonstrating all of the above when I have another minute.

**Comment by PatTheMav** [comment]
As mentioned, I cannot reproduce this on any of my machines - hotkeys work with OBS in focus and outside of it, including hotkeys consumed by the application in focus.

You are welcome to debug this locally and try to find the root cause of it. If you do, we'd be happy to get either an issue report on the root cause or (even better) a pull request to fix this. But right now I cannot reproduce this.

**Comment by jcm93** [comment]
<img width="1190" alt="Screenshot 2023-08-07 at 5 53 50 PM" src="https://github.com/obsproject/obs-studio/assets/6864788/1237a955-99f1-4c35-bb86-3f2391383661">

Inconsistent behavior: https://cdn.discordapp.com/attachments/1016400504814837811/1138259047062179920/obshotkeys7.mov

Pardon the external link; couldn't use OBS due to the breakpoint.

**Comment by jcm93** [comment]
> You are welcome to debug this locally and try to find the root cause of it. If you do, we'd be happy to get either an issue report on the root cause or (even better) a pull request to fix this. But right now I cannot reproduce this.

Yes, I am happy to work on this on my own. I would ask that the issue please be reopened in that case though, so it has a bit more visibility and others (who I know experience these symptoms) might have greater visibility on it and be able to contribute if they see it.

**Comment by Fenrirthviti** [comment]
I'm ok reopening, but added a few tags to indicate that we are not directly working on this as more info/feedback is required for action.

**Comment by ttwthomas** [comment]
could this be related to #4126 ?

**Comment by PatTheMav** [comment]
This only concerns hotkeys that need to be held down - single shot hotkeys (once that are only pressed, not held) should work as long as accessibility permissions are granted to OBS Studio.

Regarding #4126, I haven't been able to reproduce this - "normal" hotkeys work fine in 29.1.3 on all 4 Macs I regularly test with:

* Fresh hotkeys work immediately and also consecutively
* Changing hotkeys works as well
* Hotkeys work with OBS in focus as well as with another app in focus

All I need to do is copy OBS Studio into the Applications directory and give the application the required accessibility permissions, followed by a restart of the application.

**Comment by jcm93** [comment]
I don't want this to become a catch-all thread if it's not appropriate, but from the numerous videos I've posted in here, even simple single shot hotkeys are also often not working when OBS is not in focus. If we want, I can spin it out into a separate issue. It seems to be slightly different than #4126, since on my end, single shot hotkeys always work with OBS in focus, and work ~50% of the time with OBS not in focus.

Push to talk hotkeys never work, regardless of OBS focus status.

Sorry for continuing to post about this without looking into it further; I just haven't had the time to actually debug it yet.

**Comment by peppy** [comment]
To add to the conversation:

I use hotkeys to switch scenes, probably around 20-30 times over each of my dev streams (4-8 hour streams). Hotkeys always work to begin with, but enter a failing mode after some hours. But not always.

Generally it always dies in the same way:

- Global hotkeys work for a good few hours of streaming
- At some point, they stop working. When they stop, i can still focus the OBS window and use hotkeys for a bit longer (as long the windows is active).
- Eventually, even with the window focused, hotkeys stop working and I have to use the mouse to change scenes.

I pointed out in https://github.com/obsproject/obs-studio/issues/4126#issuecomment-886360920 that it may be related to secure input, but on trying to repro today I cannot.

I'm able to build/run OBS from source, so if anyone is willing to provide a branch with added debug, I can give it a run and see what things look like after they begin to fail.

FWIW I reproduced this issue across the last 4-5 major releases of macOS, on both Intel and Apple CPUs. I'm currently on an m2 studio running 13.4.1 with the latest OBS.
